pkg/web_app/lib/src/widgets.dart

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:js_interop';
+
+import 'package:web/web.dart';
+import 'package:web_app/src/completion.dart';
+
+final _widgets = <String, void Function(Element)>{
+  'completion': CompletionWidget.create,
+};
+
+void setupWidgets() {
+  for (final node in document.querySelectorAll('[data-widget]').toList()) {
+    if (!node.isA<HTMLElement>()) continue;
+    final element = node as HTMLElement;
+    final widgetName = element.getAttribute('data-widget') ?? '';
+    final widget = _widgets[widgetName];
+    try {
+      if (widget == null) {
+        throw ArgumentError('Unknown widget in data-widget="$widgetName"');
+      }
+      widget(element);
+    } catch (e, st) {
+      console.error('Failed to initialize data-widget="$widgetName"'.toJS);
+      console.error('Triggered by element:'.toJS);
+      console.error(element);
+      console.error(e.toString().toJS);
+      console.error(st.toString().toJS);
+    }
+  }
+}
+
+extension on NodeList {
+  /// Take a snapshot of [NodeList] as a Dart [List].
+  ///
+  /// Notice that it's not really safe to use [Iterable], because the underlying
```

Comment:
That's a good catch, I'm sure we are using plain `NodeList` as iterable in many places. Any idea how to enforce this better?

Reply:
Also: move them into a single place (from `completion.dart` too)

Reply:
This is from `package:web` until the rest of the JS is moved to `package:web` it doesn't really make sense.

Reply:
introducing `web_util.dart`.